# *Exploring the Influence of Weather on TTC Streetcar Delays and Forecasting Delays*

- Created on: October, 2023
- Created by: Jessica Seo

---------

## 🚂Loading Streetcar Data

### Notebook content

- Introduction
- Data Loading
- Data Cleaning
- Merging Streetcar dataset and Weather Dataset
- Data Saving


----------

### Introduction

This notebook is designed to load three individual TTC Streetcar delay datasets and merge them into a single dataset. The data covers the period from January 01, 2021, to September 30, 2023, and is collected from the City of Toronto's Open Data Source. After merging the datasets, we will create a column that can be used to join the data with the Weather dataset.

____


### Data Loading

In [1]:
#importing necessary python libraries
import pandas as pd
import time

In [2]:
#reading df1 and checking
df1 = pd.read_csv('capstone/Streetcar_Delay_2021.csv')
df1

Date Line   Time     Day             Location            Incident  \
0       1-Jan-21  501  03:15  Friday     QUEEN AND MCCAUL          Operations   
1       1-Jan-21  504  03:37  Friday  BROADVIEW AND QUEEN          Operations   
2       1-Jan-21  504  04:00  Friday    BROADVIEW STATION            Cleaning   
3       1-Jan-21  504  04:03  Friday  DUNDAS WEST STATION            Cleaning   
4       1-Jan-21  506  05:37  Friday         MAIN STATION          Mechanical   
...          ...  ...    ...     ...                  ...                 ...   
14591  31-Dec-21  510  00:39  Friday    SPADINA AND QUEEN       General Delay   
14592  31-Dec-21  504  00:51  Friday        WILSON GARAGE          Operations   
14593  31-Dec-21  509  01:19  Friday             CNE LOOP          Operations   
14594  31-Dec-21  503  01:22  Friday  QUEEN AND CONNAUGHT          Operations   
14595  31-Dec-21  505  01:53  Friday   KINGSTON AND QUEEN  Emergency Services   

       Min Delay  Min Gap Bound  Vehicle  
0             19       24     W     4574  
1             15       30   NaN     4500  
2             15       30     S     4589  
3             15       30     W     4582  
4             10       20     N     3480  
...          ...      ...   ...      ...  
14591          0        0     B        0  
14592          5       10   NaN        0  
14593          8       16   NaN     4553  
14594          8       17   NaN     4459  
14595          9       18     E     4593  

[14596 rows x 10 columns]

In [3]:
#reading df2 and checking
df2 = pd.read_csv('capstone/Streetcar_Delay_2022.csv')
df2

Date Line   Time       Day             Location  \
0       1-Jan-22  504  02:21  Saturday    BROADVIEW STATION   
1       1-Jan-22  501  03:22  Saturday    718 QUEEN ST EAST   
2       1-Jan-22  504  03:28  Saturday    BROADVIEW STATION   
3       1-Jan-22  510  03:34  Saturday        UNION STATION   
4       1-Jan-22  301  03:39  Saturday  LAKESHORE AND TENTH   
...          ...  ...    ...       ...                  ...   
17650  31-Dec-22  504  01:20  Saturday    KING AND PORTLAND   
17651  31-Dec-22  510  01:24  Saturday        UNION STATION   
17652  31-Dec-22  501  01:42  Saturday        KINGSTON LOOP   
17653  31-Dec-22  504  01:46  Saturday       KING AND PETER   
17654  31-Dec-22  504  01:47  Saturday      KING AND SIMCOE   

                       Incident  Min Delay  Min Gap Bound  Vehicle  
0      Collision - TTC Involved         30       60     E     8333  
1                    Operations         16       35     W     8068  
2                    Operations         18       36     S        0  
3                    Operations         30       60   NaN     4406  
4                      Security          5       25     W     8622  
...                         ...        ...      ...   ...      ...  
17650     Cleaning - Unsanitary         10       20     E     4588  
17651                  Security         10       20     N     4463  
17652                  Security          0        0   NaN     4534  
17653     Cleaning - Unsanitary         10       20     W     4431  
17654                 Diversion         23       33     W     4523  

[17655 rows x 10 columns]

In [4]:
#reading df3 and checking
df3 = pd.read_csv('Streetcar_Delay_2023.csv')
df3

Date Line   Time       Day                Location  \
0       1-Jan-23  509  02:37    Sunday  QUEENS QUAY AND SPADIN   
1       1-Jan-23  505  02:40    Sunday   BROADVIEW AND GERRARD   
2       1-Jan-23  504  02:52    Sunday       KING AND BATHURST   
3       1-Jan-23  504  02:59    Sunday       KING AND BATHURST   
4       1-Jan-23  509  03:33    Sunday  QUEENS QUAY AND SPADIN   
...          ...  ...    ...       ...                     ...   
10146  30-Sep-23  511  22:53  Saturday      FLEET AND MANITOBA   
10147  30-Sep-23  505  23:21  Saturday        KINGSTON RD LOOP   
10148  30-Sep-23  513  23:41  Saturday      1626 QUEEN ST EAST   
10149  30-Sep-23  501  00:48  Saturday       QUEEN AND AUGUSTA   
10150  30-Sep-23  301  01:22  Saturday       QUEEN AND AUGUSTA   

                    Incident  Min Delay  Min Gap Bound  Vehicle  
0                 Operations          0        0     E     4403  
1                    Held By         15       25     W     4460  
2      Cleaning - Unsanitary         10       20     W     4427  
3                    Held By         25       35     E     4560  
4                 Operations          0        0     E     4570  
...                      ...        ...      ...   ...      ...  
10146              Diversion         44       54     W     4593  
10147               Security         10       20   NaN     4503  
10148             Operations         10       10     W     8818  
10149              Diversion         41       61   NaN        0  
10150               Security         41       61     W     3346  

[10151 rows x 10 columns]

In [5]:
print(df1.columns)
print(df2.columns)
print(df3.columns)

Index(['Date', 'Line', 'Time', 'Day', 'Location', 'Incident', 'Min Delay',
       'Min Gap', 'Bound', 'Vehicle'],
      dtype='object')
Index(['Date', 'Line', 'Time', 'Day', 'Location', 'Incident', 'Min Delay',
       'Min Gap', 'Bound', 'Vehicle'],
      dtype='object')
Index(['Date', 'Line', 'Time', 'Day', 'Location', 'Incident', 'Min Delay',
       'Min Gap', 'Bound', 'Vehicle'],
      dtype='object')


Now that we confirmed that three dataframes has identical columns, we can confidently merge them.

In [6]:
#Merging all three dataframes and checking
df = df1.merge(df2, how='outer').merge(df3, how='outer')
df

Date Line   Time       Day             Location    Incident  \
0       1-Jan-21  501  03:15    Friday     QUEEN AND MCCAUL  Operations   
1       1-Jan-21  504  03:37    Friday  BROADVIEW AND QUEEN  Operations   
2       1-Jan-21  504  04:00    Friday    BROADVIEW STATION    Cleaning   
3       1-Jan-21  504  04:03    Friday  DUNDAS WEST STATION    Cleaning   
4       1-Jan-21  506  05:37    Friday         MAIN STATION  Mechanical   
...          ...  ...    ...       ...                  ...         ...   
42397  30-Sep-23  511  22:53  Saturday   FLEET AND MANITOBA   Diversion   
42398  30-Sep-23  505  23:21  Saturday     KINGSTON RD LOOP    Security   
42399  30-Sep-23  513  23:41  Saturday   1626 QUEEN ST EAST  Operations   
42400  30-Sep-23  501  00:48  Saturday    QUEEN AND AUGUSTA   Diversion   
42401  30-Sep-23  301  01:22  Saturday    QUEEN AND AUGUSTA    Security   

       Min Delay  Min Gap Bound  Vehicle  
0             19       24     W     4574  
1             15       30   NaN     4500  
2             15       30     S     4589  
3             15       30     W     4582  
4             10       20     N     3480  
...          ...      ...   ...      ...  
42397         44       54     W     4593  
42398         10       20   NaN     4503  
42399         10       10     W     8818  
42400         41       61   NaN        0  
42401         41       61     W     3346  

[42402 rows x 10 columns]

The data is merged well!

In [7]:
#checking the dataframe
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 42402 entries, 0 to 42401
Data columns (total 10 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   Date       42402 non-null  object
 1   Line       42186 non-null  object
 2   Time       42402 non-null  object
 3   Day        42402 non-null  object
 4   Location   42402 non-null  object
 5   Incident   42401 non-null  object
 6   Min Delay  42402 non-null  int64 
 7   Min Gap    42402 non-null  int64 
 8   Bound      35018 non-null  object
 9   Vehicle    42402 non-null  int64 
dtypes: int64(3), object(7)
memory usage: 3.6+ MB


In [8]:
#Merging Date and Time column to delay time
df["delaytime"] = pd.to_datetime(df["Date"] + " " + df["Time"])

In [9]:
#sanity check
df.head()

Date Line   Time     Day             Location    Incident  Min Delay  \
0  1-Jan-21  501  03:15  Friday     QUEEN AND MCCAUL  Operations         19   
1  1-Jan-21  504  03:37  Friday  BROADVIEW AND QUEEN  Operations         15   
2  1-Jan-21  504  04:00  Friday    BROADVIEW STATION    Cleaning         15   
3  1-Jan-21  504  04:03  Friday  DUNDAS WEST STATION    Cleaning         15   
4  1-Jan-21  506  05:37  Friday         MAIN STATION  Mechanical         10   

   Min Gap Bound  Vehicle           delaytime  
0       24     W     4574 2021-01-01 03:15:00  
1       30   NaN     4500 2021-01-01 03:37:00  
2       30     S     4589 2021-01-01 04:00:00  
3       30     W     4582 2021-01-01 04:03:00  
4       20     N     3480 2021-01-01 05:37:00

In [10]:
#Dropping Date after creating a new column
df.drop(['Date','Bound'],axis=1, inplace=True)

In [11]:
print(f'This dataframe has {df.shape[0]} rows and {df.shape[1]} columns.')

This dataframe has 42402 rows and 9 columns.


In [12]:
#sanity check
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 42402 entries, 0 to 42401
Data columns (total 9 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Line       42186 non-null  object        
 1   Time       42402 non-null  object        
 2   Day        42402 non-null  object        
 3   Location   42402 non-null  object        
 4   Incident   42401 non-null  object        
 5   Min Delay  42402 non-null  int64         
 6   Min Gap    42402 non-null  int64         
 7   Vehicle    42402 non-null  int64         
 8   delaytime  42402 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(3), object(5)
memory usage: 3.2+ MB


In [13]:
#Create a new datetime column to join Weather dataframe later
df['datetime'] =pd.to_datetime(df['delaytime'])

In [14]:
#checking
df.head()

Line   Time     Day             Location    Incident  Min Delay  Min Gap  \
0  501  03:15  Friday     QUEEN AND MCCAUL  Operations         19       24   
1  504  03:37  Friday  BROADVIEW AND QUEEN  Operations         15       30   
2  504  04:00  Friday    BROADVIEW STATION    Cleaning         15       30   
3  504  04:03  Friday  DUNDAS WEST STATION    Cleaning         15       30   
4  506  05:37  Friday         MAIN STATION  Mechanical         10       20   

   Vehicle           delaytime            datetime  
0     4574 2021-01-01 03:15:00 2021-01-01 03:15:00  
1     4500 2021-01-01 03:37:00 2021-01-01 03:37:00  
2     4589 2021-01-01 04:00:00 2021-01-01 04:00:00  
3     4582 2021-01-01 04:03:00 2021-01-01 04:03:00  
4     3480 2021-01-01 05:37:00 2021-01-01 05:37:00

In [15]:
#Replacing the minute to 0 in datetime column to match with weather dataframe later
df['datetime']= df['datetime'].apply(lambda x:x.replace(minute=0))

In [16]:
#checking
df.head(3)

Line   Time     Day             Location    Incident  Min Delay  Min Gap  \
0  501  03:15  Friday     QUEEN AND MCCAUL  Operations         19       24   
1  504  03:37  Friday  BROADVIEW AND QUEEN  Operations         15       30   
2  504  04:00  Friday    BROADVIEW STATION    Cleaning         15       30   

   Vehicle           delaytime            datetime  
0     4574 2021-01-01 03:15:00 2021-01-01 03:00:00  
1     4500 2021-01-01 03:37:00 2021-01-01 03:00:00  
2     4589 2021-01-01 04:00:00 2021-01-01 04:00:00

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 42402 entries, 0 to 42401
Data columns (total 10 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Line       42186 non-null  object        
 1   Time       42402 non-null  object        
 2   Day        42402 non-null  object        
 3   Location   42402 non-null  object        
 4   Incident   42401 non-null  object        
 5   Min Delay  42402 non-null  int64         
 6   Min Gap    42402 non-null  int64         
 7   Vehicle    42402 non-null  int64         
 8   delaytime  42402 non-null  datetime64[ns]
 9   datetime   42402 non-null  datetime64[ns]
dtypes: datetime64[ns](2), int64(3), object(5)
memory usage: 3.6+ MB


------
### Merging Streetcar data and Weather data

In [18]:
#Loading Weather data
weather = pd.read_csv('final_weather.csv')
weather

Temp Definition °C Dew Point Definition °C Rel Hum Definition %  \
0                   -1.3                    -4.7                   78   
1                   -1.2                    -3.6                   84   
2                   -1.8                    -3.2                   90   
3                   -2.0                    -3.3                   91   
4                   -1.4                    -3.5                   85   
...                  ...                     ...                  ...   
24067               17.6                    15.7                   89   
24068               17.0                    15.7                   92   
24069               16.6                    15.1                   91   
24070               16.6                    15.1                   91   
24071               16.9                    16.3                   96   

       Precip. Amount Definition mm Wind Dir Definition 10's deg  \
0                               0.0                         25.0   
1                               0.0                         24.0   
2                               0.0                         27.0   
3                               0.0                         30.0   
4                               0.0                         27.0   
...                             ...                          ...   
24067                           0.0                           24   
24068                           0.0                           25   
24069                           0.0                           26   
24070                           0.0                     LegendMM   
24071                           0.0                     LegendMM   

      Wind Spd Definition km/h Visibility Definition km  \
0                            4                     16.1   
1                            4                     16.1   
2                            4                     16.1   
3                            5                     16.1   
4                            5                     16.1   
...                        ...                      ...   
24067                        4                     16.1   
24068                        5                     16.1   
24069                        8                     16.1   
24070                        4                     16.1   
24071                        4                     16.1   

      Stn Press Definition kPa  Hmdx Definition  Wind Chill Definition  \
0                       102.17              NaN                   -3.0   
1                       102.16              NaN                   -3.0   
2                       102.19              NaN                   -3.0   
3                       102.26              NaN                   -4.0   
4                       102.24              NaN                   -3.0   
...                        ...              ...                    ...   
24067                   101.52              NaN                    NaN   
24068                   101.53              NaN                    NaN   
24069                   101.53              NaN                    NaN   
24070                   101.57              NaN                    NaN   
24071                   101.59              NaN                    NaN   

      Weather Definition             datetime  
0             LegendNANA  2021-01-01 00:00:00  
1             LegendNANA  2021-01-01 01:00:00  
2             LegendNANA  2021-01-01 02:00:00  
3             LegendNANA  2021-01-01 03:00:00  
4             LegendNANA  2021-01-01 04:00:00  
...                  ...                  ...  
24067         LegendNANA  2023-09-30 19:00:00  
24068         LegendNANA  2023-09-30 20:00:00  
24069         LegendNANA  2023-09-30 21:00:00  
24070         LegendNANA  2023-09-30 22:00:00  
24071         LegendNANA  2023-09-30 23:00:00  

[24072 rows x 12 columns]

In [19]:
weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24072 entries, 0 to 24071
Data columns (total 12 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Temp Definition °C            24044 non-null  object 
 1   Dew Point Definition °C       24044 non-null  object 
 2   Rel Hum Definition %          24044 non-null  object 
 3   Precip. Amount Definition mm  24044 non-null  float64
 4   Wind Dir Definition 10's deg  23632 non-null  object 
 5   Wind Spd Definition km/h      24044 non-null  object 
 6   Visibility Definition km      24044 non-null  object 
 7   Stn Press Definition kPa      24044 non-null  object 
 8   Hmdx Definition               4005 non-null   float64
 9   Wind Chill Definition         3885 non-null   float64
 10  Weather Definition            24044 non-null  object 
 11  datetime                      24072 non-null  object 
dtypes: float64(3), object(9)
memory usage: 2.2+ MB


In [20]:
#converting datetime to datetime dtype to merge
weather["datetime"] = pd.to_datetime(weather['datetime'])

Explain why you are dropping them

In [21]:
#dropping five irrelevent columns
drops = ['Dew Point Definition °C','Rel Hum Definition %','Wind Dir Definition 10\'s deg','Wind Spd Definition km/h','Stn Press Definition kPa']

weather=weather.drop(drops, axis =1)

In [22]:
weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 24072 entries, 0 to 24071
Data columns (total 7 columns):
 #   Column                        Non-Null Count  Dtype         
---  ------                        --------------  -----         
 0   Temp Definition °C            24044 non-null  object        
 1   Precip. Amount Definition mm  24044 non-null  float64       
 2   Visibility Definition km      24044 non-null  object        
 3   Hmdx Definition               4005 non-null   float64       
 4   Wind Chill Definition         3885 non-null   float64       
 5   Weather Definition            24044 non-null  object        
 6   datetime                      24072 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(3), object(3)
memory usage: 1.3+ MB


In [23]:
weather.head()

Temp Definition °C  Precip. Amount Definition mm Visibility Definition km  \
0               -1.3                           0.0                     16.1   
1               -1.2                           0.0                     16.1   
2               -1.8                           0.0                     16.1   
3               -2.0                           0.0                     16.1   
4               -1.4                           0.0                     16.1   

   Hmdx Definition  Wind Chill Definition Weather Definition  \
0              NaN                   -3.0         LegendNANA   
1              NaN                   -3.0         LegendNANA   
2              NaN                   -3.0         LegendNANA   
3              NaN                   -4.0         LegendNANA   
4              NaN                   -3.0         LegendNANA   

             datetime  
0 2021-01-01 00:00:00  
1 2021-01-01 01:00:00  
2 2021-01-01 02:00:00  
3 2021-01-01 03:00:00  
4 2021-01-01 04:00:00

In [24]:
print(f'This dataframe has {weather.shape[0]} rows and {weather.shape[1]} columns.')

This dataframe has 24072 rows and 7 columns.


In [25]:
#testing to merge before merging the whole datasets
streetcar_test=df.head(20)
weather_test=weather.head(20)

In [26]:
#testing the merge
test_test=streetcar_test.merge(weather_test, on='datetime', how='left')
test_test

Line   Time     Day                Location            Incident  Min Delay  \
0   501  03:15  Friday        QUEEN AND MCCAUL          Operations         19   
1   504  03:37  Friday     BROADVIEW AND QUEEN          Operations         15   
2   504  04:00  Friday       BROADVIEW STATION            Cleaning         15   
3   504  04:03  Friday     DUNDAS WEST STATION            Cleaning         15   
4   506  05:37  Friday            MAIN STATION          Mechanical         10   
5   555  06:00  Friday  TORONTO TRANSIT COMMIS       General Delay          0   
6   501  06:59  Friday            RUSSELL YARD  Emergency Services         10   
7   504  07:55  Friday          KING AND QUEEN          Mechanical         10   
8   511  09:35  Friday      BATHURST AND QUEEN            Cleaning          8   
9   512  09:55  Friday        ST CLAIR STATION  Emergency Services          7   
10  512  10:01  Friday        ST CLAIR STATION             Held By          6   
11  504  10:28  Friday          KING AND QUEEN          Mechanical          8   
12  510  11:07  Friday       SPADINA AND QUEEN          Operations          7   
13  511  11:51  Friday   ST CLAIR AND DUFFERIN  Utilized Off Route          5   
14  510  11:59  Friday  SPADINA AND QUEENS QUA                Late          7   
15  510  12:00  Friday          CHARLOTTE LOOP                Late          7   
16  506  12:00  Friday     DUNDAS WEST STATION  Utilized Off Route          7   
17  510  12:02  Friday         SPADINA STATION                Late          7   
18  510  12:13  Friday        SPADINA AND KING                Late          7   
19  506  12:28  Friday     MAIN STREET STATION          Operations          7   

    Min Gap  Vehicle           delaytime            datetime  \
0        24     4574 2021-01-01 03:15:00 2021-01-01 03:00:00   
1        30     4500 2021-01-01 03:37:00 2021-01-01 03:00:00   
2        30     4589 2021-01-01 04:00:00 2021-01-01 04:00:00   
3        30     4582 2021-01-01 04:03:00 2021-01-01 04:00:00   
4        20     3480 2021-01-01 05:37:00 2021-01-01 05:00:00   
5         0        0 2021-01-01 06:00:00 2021-01-01 06:00:00   
6        20     4496 2021-01-01 06:59:00 2021-01-01 06:00:00   
7        20     4520 2021-01-01 07:55:00 2021-01-01 07:00:00   
8        16     1406 2021-01-01 09:35:00 2021-01-01 09:00:00   
9        12     4543 2021-01-01 09:55:00 2021-01-01 09:00:00   
10       12     4532 2021-01-01 10:01:00 2021-01-01 10:00:00   
11       16     4491 2021-01-01 10:28:00 2021-01-01 10:00:00   
12       14     4555 2021-01-01 11:07:00 2021-01-01 11:00:00   
13       10     1036 2021-01-01 11:51:00 2021-01-01 11:00:00   
14       14     4436 2021-01-01 11:59:00 2021-01-01 11:00:00   
15       14     4556 2021-01-01 12:00:00 2021-01-01 12:00:00   
16       14     1037 2021-01-01 12:00:00 2021-01-01 12:00:00   
17       14     4502 2021-01-01 12:02:00 2021-01-01 12:00:00   
18       14     4577 2021-01-01 12:13:00 2021-01-01 12:00:00   
19       14     1101 2021-01-01 12:28:00 2021-01-01 12:00:00   

   Temp Definition °C  Precip. Amount Definition mm Visibility Definition km  \
0                -2.0                           0.0                     16.1   
1                -2.0                           0.0                     16.1   
2                -1.4                           0.0                     16.1   
3                -1.4                           0.0                     16.1   
4                 0.4                           0.0                     16.1   
5                 0.0                           0.0                     16.1   
6                 0.0                           0.0                     16.1   
7                 0.5                           0.0                     16.1   
8                 1.4                           0.0                     16.1   
9                 1.4                           0.0                     16.1   
10                1.5                           0.0                     16.1

Don't think delaytime column is needed, so make the decision to drop after merging everything.

It seems to have merged correctly. We will continue to merge all the rows.

In [27]:
#Finally merging the whole datasets after checking that it went well.
final_dataframe=df.merge(weather, on='datetime', how='left')

In [28]:
#checking 
final_dataframe

Line   Time       Day             Location    Incident  Min Delay  \
0      501  03:15    Friday     QUEEN AND MCCAUL  Operations         19   
1      504  03:37    Friday  BROADVIEW AND QUEEN  Operations         15   
2      504  04:00    Friday    BROADVIEW STATION    Cleaning         15   
3      504  04:03    Friday  DUNDAS WEST STATION    Cleaning         15   
4      506  05:37    Friday         MAIN STATION  Mechanical         10   
...    ...    ...       ...                  ...         ...        ...   
42397  511  22:53  Saturday   FLEET AND MANITOBA   Diversion         44   
42398  505  23:21  Saturday     KINGSTON RD LOOP    Security         10   
42399  513  23:41  Saturday   1626 QUEEN ST EAST  Operations         10   
42400  501  00:48  Saturday    QUEEN AND AUGUSTA   Diversion         41   
42401  301  01:22  Saturday    QUEEN AND AUGUSTA    Security         41   

       Min Gap  Vehicle           delaytime            datetime  \
0           24     4574 2021-01-01 03:15:00 2021-01-01 03:00:00   
1           30     4500 2021-01-01 03:37:00 2021-01-01 03:00:00   
2           30     4589 2021-01-01 04:00:00 2021-01-01 04:00:00   
3           30     4582 2021-01-01 04:03:00 2021-01-01 04:00:00   
4           20     3480 2021-01-01 05:37:00 2021-01-01 05:00:00   
...        ...      ...                 ...                 ...   
42397       54     4593 2023-09-30 22:53:00 2023-09-30 22:00:00   
42398       20     4503 2023-09-30 23:21:00 2023-09-30 23:00:00   
42399       10     8818 2023-09-30 23:41:00 2023-09-30 23:00:00   
42400       61        0 2023-09-30 00:48:00 2023-09-30 00:00:00   
42401       61     3346 2023-09-30 01:22:00 2023-09-30 01:00:00   

      Temp Definition °C  Precip. Amount Definition mm  \
0                   -2.0                           0.0   
1                   -2.0                           0.0   
2                   -1.4                           0.0   
3                   -1.4                           0.0   
4                    0.4                           0.0   
...                  ...                           ...   
42397               16.6                           0.0   
42398               16.9                           0.0   
42399               16.9                           0.0   
42400               16.9                           0.0   
42401               16.6                           0.0   

      Visibility Definition km  Hmdx Definition  Wind Chill Definition  \
0                         16.1              NaN                   -4.0   
1                         16.1              NaN                   -4.0   
2                         16.1              NaN                   -3.0   
3                         16.1              NaN                   -3.0   
4                         16.1              NaN                    NaN   
...                        ...              ...                    ...   
42397                     16.1              NaN                    NaN   
42398                     16.1              NaN                    NaN   
42399                     16.1              NaN                    NaN   
42400                     16.1              NaN                    NaN   
42401                     12.9              NaN                    NaN   

      Weather Definition  
0             LegendNANA  
1             LegendNANA  
2             LegendNANA  
3             LegendNANA  
4             LegendNANA  
...                  ...  
42397         LegendNANA  
42398         LegendNANA  
42399         LegendNANA  
42400         LegendNANA  
42401         LegendNANA  

[42402 rows x 16 columns]

In [29]:
final_dataframe=final_dataframe.drop('delaytime', axis =1)

In [30]:
final_dataframe.head()

Line   Time     Day             Location    Incident  Min Delay  Min Gap  \
0  501  03:15  Friday     QUEEN AND MCCAUL  Operations         19       24   
1  504  03:37  Friday  BROADVIEW AND QUEEN  Operations         15       30   
2  504  04:00  Friday    BROADVIEW STATION    Cleaning         15       30   
3  504  04:03  Friday  DUNDAS WEST STATION    Cleaning         15       30   
4  506  05:37  Friday         MAIN STATION  Mechanical         10       20   

   Vehicle            datetime Temp Definition °C  \
0     4574 2021-01-01 03:00:00               -2.0   
1     4500 2021-01-01 03:00:00               -2.0   
2     4589 2021-01-01 04:00:00               -1.4   
3     4582 2021-01-01 04:00:00               -1.4   
4     3480 2021-01-01 05:00:00                0.4   

   Precip. Amount Definition mm Visibility Definition km  Hmdx Definition  \
0                           0.0                     16.1              NaN   
1                           0.0                     16.1              NaN   
2                           0.0                     16.1              NaN   
3                           0.0                     16.1              NaN   
4                           0.0                     16.1              NaN   

   Wind Chill Definition Weather Definition  
0                   -4.0         LegendNANA  
1                   -4.0         LegendNANA  
2                   -3.0         LegendNANA  
3                   -3.0         LegendNANA  
4                    NaN         LegendNANA

The dataset's time frame is from January 01, 2021 to September 30, 2023. Looking at the above dataframe tell us that everything is loaded correctly. Streetcar data has 42402 rows and 10 columns and Weather data has 24072 rows and 12 columns. It make sense that the new dataframe has 42402 rows and 21 columns.

In [31]:
final_dataframe.sample(20)

Line   Time        Day                Location            Incident  \
33049  510  10:10  Wednesday      SPADINA AND DUNDAS            Security   
1900   510  13:45   Saturday  SPADINA AND QUEEN'S QU            Cleaning   
29914  501  08:30   Thursday     1262 LAWRENCE AVE W          Operations   
22263  511  23:13     Sunday          CHARLOTTE LOOP            Security   
255    511  18:41  Wednesday        BATHURST STATION       Investigation   
14616  506  09:04   Saturday  HOWARD PARK AND RONCES          Operations   
2839   505  01:35    Tuesday   DUNDAS AND SHERBOURNE            Cleaning   
2516   506  16:18    Tuesday       GERRARD AND JONES            Cleaning   
23340  504  21:59     Sunday  RONCESVALLES AND DUNDA           Diversion   
38226  504  01:55   Saturday       KING AND DUFFERIN  Emergency Services   
52     505  05:22   Saturday        DUNDAS AND YONGE            Security   
27665  512  16:24  Wednesday   ST CLAIR AND BATHURST       General Delay   
36243  505  23:42   Saturday  ERINDALE AND BROADVIEW           Diversion   
22622  506  08:38   Saturday          HIGH PARK LOOP          Operations   
4170   506  17:54     Monday   GERRARD AND GREENWOOD          Operations   
33494  304  01:38   Saturday       KING AND BATHURST           Diversion   
6187   504  04:36     Sunday     DUNDAS WEST STATION          Operations   
24795  504  18:04   Thursday        DUNDAS AND BLOOR          Operations   
41139  501  21:48   Thursday      QUEEN AND SORAUREN           Diversion   
6744   504  12:32     Monday              FLEET LOOP          Operations   

       Min Delay  Min Gap  Vehicle            datetime Temp Definition °C  \
33049         10       10     4457 2023-01-25 10:00:00                0.3   
1900           4        8     4502 2021-02-27 13:00:00                5.6   
29914          6       12     8372 2022-11-03 08:00:00                7.8   
22263         10       20     4488 2022-05-08 23:00:00               10.8   
255            7       15     4505 2021-01-06 18:00:00                2.0   
14616          9       18        0 2022-01-01 09:00:00                4.8   
2839          10       20     4488 2021-03-23 01:00:00                8.2   
2516           6       12     3418 2021-03-16 16:00:00                1.7   
23340        162      169     8332 2022-05-29 21:00:00           LegendMM   
38226         24       34     4582 2023-06-03 01:00:00               24.4   
52            16       27     4422 2021-01-02 05:00:00                0.1   
27665         10       15     4540 2022-09-07 16:00:00               21.0   
36243         45       55     4503 2023-04-15 23:00:00               12.0   
22622          9       18        0 2022-05-14 08:00:00               16.9   
4170           0        0     3178 2021-04-26 17:00:00                5.9   
33494         60       70     3519 2023-02-04 01:00:00              -19.9   
6187           8       16        0 2021-06-20 04:00:00               13.5   
24795          5       10     8111 2022-06-30 18:00:00               20.2   
41139         32       40     4492 2023-08-24 21:00:00               18.8   
6744           5       10     4549 2021-07-05 12:00:00               22.9   

       Precip. Amount Definition mm Visibility Definition km  Hmdx Definition  \
33049                           0.0                      9.7              NaN   
1900                            0.0                     16.1              NaN   
29914                           0.0                      0.6              NaN   
22263                           0.0                     16.1              NaN   
255                             0.0                     16.1              NaN   
14616                           0.0                     16.1              NaN   
2839                            0.0                     16.1              NaN   
2516                            0.0                     16.1              NaN   
23340                           0.0                     16.1  

----
###  Data Saving

In [32]:
#saving the dataframe to a new csv file! 
final_dataframe.to_csv('Streetcar_Weather_Data.csv', index=False)